pkg/controller/dnszone/dnszone_controller.go

```diff
@@ -199,13 +172,102 @@ func (r *ReconcileDNSZone) Reconcile(request reconcile.Request) (reconcile.Resul
 		"currentGeneration":  desiredState.Generation,
 		"lastSyncGeneration": desiredState.Status.LastSyncGeneration,
 	}).Info("Syncing DNS Zone")
-	result, err := zr.Reconcile()
+	result, err := r.ReconcileDNSProvider(actuator, desiredState)
 	if err != nil {
 		dnsLog.WithError(err).Error("Encountered error while attempting to reconcile")
 	}
 	return result, err
 }
 
+// ReconcileDNSProvider attempts to make the current state reflect the desired state. It does this idempotently.
+func (r *ReconcileDNSZone) ReconcileDNSProvider(actuator Actuator, dnsZone *hivev1.DNSZone) (reconcile.Result, error) {
+	r.logger.Debug("Retrieving current state")
+	err := actuator.Refresh()
```

Comment:
`Refresh` is sticking out to me as something that is not well defined. For one thing, why does refresh fetch the zone but not the name servers. For another thing, it smells to me that `Refresh` is something that needs to be called immediately after creating the actuator and before doing anything else.

I have two alternatives in mind.
1) Have the `Exists` function fetch the zone ID and store it.
2) Have the `NewAWSActuator` function fetch the zone ID and store it. This is my preference.

In either case, there is no reason to store the tags: They can be fetched when needed in the `syncTags` function.

Reply:
I have the other changes done, but I don't fully understand what you're looking for on this one.

Are you looking to eliminate `Refresh()`?

What is the purpose of fetching and storing zoneID in `NewAWSActuator`? Like what does it accomplish?

The `syncTags` part is clear.

Reply:
Yes, my suggestions is to eliminate `Refresh`. My issue is that `NewAWSActuator` does not yield anything usable. The actuator created is not usable until `Refresh` is called. If the user is required to call `Refresh` immediately after calling `NewAWSActuator`, then why not include the functionality of `Refresh` in the `NewAWSActuator` function and skip the extra function call.

Reply:
Ok. Why store `zoneID` instead of `currentHostedZone`?

Would we just look up the hosted zone each time?

Is this to ensure the hosted zone isn't "stale" ?

Reply:
I don't care much whether the actuator stores the entire hosted zone or just the ID. For what it's worth, though, nothing in the hosted zone other than the ID is used in the actuator.

Reply:
You're right! I hadn't noticed that. :)

Reply:
So I've been thinking about this change this weekend and while I see your point on Refresh, I'm just not convinced that putting that functionality into `NewAWSActuator` is a good idea.

1. `NewAWSActuator` is basically a constructor. I don't like the idea of having the constructor call out to the cloud.
1. I was also hoping to not significantly alter the flow control / logic of the current dnszone controller for the actuator refactor. Adopting the existing logic gives us the highest chance of not introducing bugs.
1. I'd like more time to think of a better way to address this concern. But I don't think we can afford to take that time right now. If we want to get the GCP actuator done this sprint, I really need to start focusing on it.

Proposal: I'll create a Jira card to think of and implement something to address this concern. Between now and when we do that card I'll think about a way to address the concern without having the constructor reach out to the cloud. To be clear, this card would also handle the `syncTags` part as well since the two seem related.

I have the other feedback that you've pointed out addressed (except putting the re-vendor of the SOA check deps into its own commit, which I can do today). I haven't pushed any of those changes yet.

@staebler @dgoodwin thoughts?

Reply:
1. I agree about getting this in without making any more changes around `Refresh` at this point.
2. Yes, `NewAWSActuator` is a constructor. An object returned by a constructor should be usable. This constructor does not return a usable object until `Refresh` is called. That is my issue.
3. I do not see where putting the `Refresh` logic in a stand-alone function or in the constructor changes the flow/logic in any way. I could make an argument that adding the `NewAWSActuator` as an additional function while having `Refresh` replace `getCurrentState` actually makes more changes than having `NewAWSActuator` be a replacement for `getCurrentState`.
4. If the concern is that it is peculiar to have the constructor make calls out to the cloud, then you could have each of the public functions that requires a refresh in order to be usable (1) check whether a refresh has been executed and (2) execute the refresh if needed.
5. I would not bother created a Jira card. I doubt that the card would ever be worked on. Once this PR merges, that is going to be the interface for the Actuator. And this is fine: It is just an internal interface used by one component so it does not need to be perfect.